middleware/ratelimiter.go

```diff
@@ -0,0 +1,511 @@
+			Strategy:"ip",
+			Key:"",
+		},
+		Prefix:"LIMIT",
+		Client:nil,
+		SkipRateLimiterInternalError:false,
+	}
+	limiterImp *limiter
+)
+
+// RateLimiter returns a rate limit middleware.
+func RateLimiter() echo.MiddlewareFunc {
+	return RateLimiterWithConfig(DefaultRateLimiterConfig)
+}
+
+// RateLimiterWithConfig returns a RateLimiter middleware with config.
+// See: `RateLimiter()`.
+func RateLimiterWithConfig(config RateLimiterConfig) echo.MiddlewareFunc {
+	// Defaults
+	if config.Skipper == nil {
+		config.Skipper = DefaultRateLimiterConfig.Skipper
+	}
+	if len(config.LimitConfig.Strategy) == 0 {
+		panic("echo: rate limiter middleware requires Strategy (use ip or header(key)")
+	}
+	if config.Prefix == "" {
+		config.Prefix = "LIMIT:"
+	}
+	if config.LimitConfig.Max <= 0 {
+		config.LimitConfig.Max = 100
+	}
+	if config.LimitConfig.Duration <= 0 {
+		config.LimitConfig.Duration = time.Minute *1
+	}
+
+	//If config.Client omit, the limiter is a memory limiter
+	if config.Client == nil {
+		limiterImp = newMemoryLimiter(&config)
+	}else{
+		//setup redis client
+		limiterImp = newRedisLimiter(&config)
+	}
+
+	var tokenExtractor TokenExtractor
+
+	switch strings.ToLower(config.LimitConfig.Strategy) {
+	case "ip":
+		tokenExtractor = IPTokenExtractor()
+	case "header":
+		tokenExtractor = HeaderTokenExtractor(config.LimitConfig.Key)
+	}
+
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+			response := c.Response()
+
+			policy := []int{}
+			result, err := limiterImp.Get(tokenExtractor(c),policy...)
+
+			if err != nil {
+
+				if config.SkipRateLimiterInternalError{
+					return next(c)
+				}
+				return echo.NewHTTPError(http.StatusBadRequest, err.Error())
+			}
+
+			response.Header().Set("X-Ratelimit-Limit", strconv.FormatInt(int64(result.Total), 10))
+			response.Header().Set("X-Ratelimit-Remaining", strconv.FormatInt(int64(result.Remaining), 10))
+			response.Header().Set("X-Ratelimit-Reset", strconv.FormatInt(result.Reset.Unix(), 10))
+
+			if result.Remaining <= 0 {
+
+				after := int64(result.Reset.Sub(time.Now())) / 1e9
+				response.Header().Set("Retry-After", strconv.FormatInt(after, 10))
+				return echo.NewHTTPError(http.StatusTooManyRequests, "Rate limit exceeded, retry in %d seconds.\n", after)
+			}
+			return next(c)
+		}
+	}
+}
+
+
+// TokenExtractor defines the interface of the functions to use in order to extract a token for each request
+type TokenExtractor func(echo.Context) string
+
+// IPTokenExtractor extracts the IP of the request
+func IPTokenExtractor() TokenExtractor {
+	return func(c echo.Context) string { return strings.Split(c.RealIP(), ":")[0]  }
+}
+
+// HeaderTokenExtractor returns a TokenExtractor that looks for the value of the designed header
+func HeaderTokenExtractor(header string) TokenExtractor {
+	return func(c echo.Context) string { return c.Request().Header.Get(header) }
+}
```

Comment:
This function should be injectable, e.g. for `x-forwarded-for` when you only want to use the right-most value.